subprojects/language-java/src/integTest/groovy/org/gradle/language/java/JavaLanguageIncrementalBuildIntegrationTest.groovy

```diff
@@ -38,4 +38,33 @@ class JavaLanguageIncrementalBuildIntegrationTest extends AbstractJvmLanguageInc
         executedAndNotSkipped ":compileMainJarMainJava", ":createMainJar", ":mainJar"
     }
 
-}
\ No newline at end of file
+    def "task outcome is up to date when no recompilation necessary"() {
+        given:
+        buildFile << '''
+            tasks.withType(org.gradle.api.tasks.compile.AbstractCompile) { task ->
+                task.options.incremental = true
+                task.inputs.file 'input.txt' 
+            }
+        '''.stripIndent()
+        file('input.txt') << 'first run'
+
+        when:
+        succeeds mainCompileTaskName
+
+        then:
+        executedAndNotSkipped mainCompileTaskName
+
+        when:
+        file('input.txt') << 'second run, triggers task execution, but no recompilation is necessary'
+
+        and:
+        succeeds mainCompileTaskName, '--debug'
+
+        then:
+        executed mainCompileTaskName
```

Comment:
Checking only for skipped should be enough - it it has been skipped it also has been executed.